util/chunk/chunk.go

```diff
@@ -30,24 +30,76 @@ type Chunk struct {
 	// numVirtualRows indicates the number of virtual rows, which have zero column.
 	// It is used only when this Chunk doesn't hold any data, i.e. "len(columns)==0".
 	numVirtualRows int
+	maxRows        int
+	maxCap         int
 }
 
 // Capacity constants.
 const (
-	InitialCapacity = 32
+	InitialCapacity    = 32
+	VoidCapacity       = -1
+	PerChunkSizeInList = 1024
 )
 
-// NewChunkWithCapacity creates a new chunk with field types and capacity.
-func NewChunkWithCapacity(fields []*types.FieldType, cap int) *Chunk {
+// NewFixedChunk creates a new chunk with field types and capacity.
+func NewFixedChunk(fields []*types.FieldType, cap, maxCap int) *Chunk {
 	chk := new(Chunk)
+	// void capacity will return a void chunk without allocate columns.
+	if cap == VoidCapacity {
+		return chk
+	}
 	chk.columns = make([]*column, 0, len(fields))
 	chk.numVirtualRows = 0
 	for _, f := range fields {
 		chk.addColumnByFieldType(f, cap)
 	}
+	chk.maxCap = maxCap
+	chk.maxRows = cap
 	return chk
 }
 
+// Renew creates a new chunk with double capacity if full.
+func (c *Chunk) Renew(fields []*types.FieldType) *Chunk {
+	if c.NumRows() < c.maxRows || c.NumRows() == c.maxCap {
+		return NewFixedChunk(fields, c.NumRows(), c.maxCap)
+	}
+	c.maxRows = c.maxRows * 2
+	if c.maxRows > c.maxCap {
+		c.maxRows = c.maxCap
+	}
+	return NewFixedChunk(fields, c.maxRows, c.maxCap)
+}
+
+// NewSize computes the next pull chunk capacity.
+// Only used for want to rename but could not get retTypes()
+func (c *Chunk) NewSize() int {
+	if c.NumRows() < c.maxRows || c.NumRows() == c.maxCap {
+		return c.NumRows()
+	}
+	c.maxRows = c.maxRows * 2
+	if c.maxRows > c.maxCap {
+		c.maxRows = c.maxCap
+	}
+	return c.maxRows
+}
+
+// GrowIfFull grows the capacity be reallocate chunk if previous pull if full.
+func (c *Chunk) GrowIfFull() {
+	if c.NumRows() < c.maxRows || c.NumRows() == c.maxCap {
+		return
+	}
+	lastNumRows := c.NumRows()
+	c.maxRows = c.maxRows * 2
+	if c.maxRows > c.maxCap {
+		c.maxRows = c.maxCap
+	}
+	c.numVirtualRows = 0
+	for _, col := range c.columns {
+		col.grow(c.maxRows, lastNumRows)
```

Comment:
Isn't the `lastNumRows` is the same as `col.length`?

Reply:
it's same i will fix it